### PR TITLE
Implement conditional types slightly better

### PR DESCRIPTION
**Pull request by rvanvelzen**
Still a proof of concept, but possibly slightly more useful than #1120 for phpstan/phpstan#3853

### CODE REVIEW

**Comment by ondrejmirtes** [comment]
I guess this is gonna work well. I was first a bit sceptical about this because it burdens the type system with information that isn't relevant for types, like parameter names, but on the other hand, I don't see a different way to make conditional types in nested normal types work.

Maybe we could even make the ConditionalType* not extend UnionType (because there can be a situation where the if/else branches actually don't lead to `UnionType` after normalization (https://phpstan.org/developing-extensions/type-system#type-normalization). And transform the type to something useful in similar places where StaticType is transformed using TypeTraverser.

With https://github.com/phpstan/phpstan-src/pull/1133 reverted, you can also forget about these todos from https://github.com/phpstan/phpstan-src/pull/1120#issuecomment-1080516416:

* Easy: PhpDocNodeResolver: add new method resolveConditionalReturnTag(): ConditionalReturnTypeTag|ConditionalReturnTypeForParameterTag|null
* Easy: ResolvedPhpDocBlock: add new method getConditionalReturnTag(): ConditionalReturnTypeTag|ConditionalReturnTypeForParameterTag|null
* Hard: Add new interface ParametersAcceptorWithConditionalReturnTag. Go through the implementations of ParametersAcceptor and decide where it should be implemented. I suspect that FunctionVariant and FunctionVariantWithPhpDocs will gain a new optional constructor parameter with the conditional return type tag.
* Hard: Go through the places where FunctionVariant and FunctionVariantWithPhpDocs are instantiated and try to pass the conditional return type tag there if it makes sense.

These complications are no longer needed. With the current PR, you can go straight to GenericParametersAcceptorResolver and try to implement the conditional types logic :) Because the conditional type is present in `$parametersAcceptor->getReturnType()`.

File: src/Type/ConditionalType.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type;
+
+use PHPStan\Type\Generic\TemplateType;
+use PHPStan\Type\Generic\TemplateTypeVariance;
+use function sprintf;
+
+/** @api */
+final class ConditionalType extends UnionType
```
Comment (ondrejmirtes) [inline src/Type/ConditionalType.php]: An idea - do not implement UnionType, just Type, create `TypeCombinator::union($if, $else)` in the constructor and delegate all Type method calls to that type.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Type/ConditionalType.php (hunk 1 above)
Comment (ZebulanStanphill) [reply src/Type/ConditionalType.php]: Yeah, I think this is definitely a case where the classic principle of "prefer composition over inheritance" applies.

Comment (rvanvelzen) [reply src/Type/ConditionalType.php]: Sure. For prototyping it was very useful though. :)

Comment (ondrejmirtes) [reply src/Type/ConditionalType.php]: Awesome! I love the progress, seems very promising :)

File: src/Reflection/ResolvedFunctionVariant.php [hunk 2]
```diff
@@ -73,10 +81,29 @@ public function getReturnType(): Type
...
+				if (array_key_exists($type->getParameterName(), $this->passedArgs)) {
+					$type = $type->toConditional($this->passedArgs[$type->getParameterName()]);
+				}
+			}
+
+			if ($type instanceof ConditionalType) {
+				$type = $type->resolve();
+			}
+
+			return $traverse($type);
```
Comment (ondrejmirtes) [inline src/Reflection/ResolvedFunctionVariant.php:98]: This looks like it supports nested conditionals but they're not tested :) Please add some test conditional type inside conditional type :) Thanks.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Also some tests for NodeScopeResolverTest how conditionals (that are wrongly configured) behave would be nice :)

Examples: ConditionalType with non-template type in $subject, ConditionalTypeForParameter with non-existent parameter name.

Otherwise it's great and I'm gonna merge this :)

**Comment by herndlm** [comment]
It's great watching this coming together in small steps 👍

**Comment by ondrejmirtes** [comment]
Added an up-to-date checklist to the issue: https://github.com/phpstan/phpstan/issues/3853#issuecomment-1082351763

**Comment by ondrejmirtes** [comment]
Thank you, this is awesome. I just have one question - why is there https://github.com/phpstan/phpstan-src/blob/e0e7ba45f923691c5a8a6b1873e357113b62e6eb/src/Reflection/ResolvedFunctionVariant.php#L93-L100 a while loop? Under what condition is `ConditionalTypeForParameter` gonna stay `ConditionalTypeForParameter` after calling `toConditional`?

**Comment by rvanvelzen** [comment]
`toConditional()` substitutes the parameter for a `Type` and returns the result of `Conditional::create()`. If that is resolvable immediately, it will return the appropriate result. That result can then also be a `ConditionalTypeForParameter`. `tests/PHPStan/Analyser/data/conditional-types.php:57` covers this case.

**Comment by ondrejmirtes** [comment]
Yeah, I realized this morning it's for the nested conditionals :) Thanks. I really appreciate your work here (and I have an idea for another similar feature once we're done with this one :))

**Pull request merged**